src/rulesets/Base/array.jl

```diff
@@ -43,32 +51,81 @@ function rrule(::typeof(Base.vect), X::Vararg{Any,N}) where {N}
     return Base.vect(X...), vect_pullback
 end
 
+"""
+    _make_real_zeros(xdots, xs)
+
+Forward rules for `vect` or `cat` may receive a mixture of data and `ZeroTangent`s.
+To avoid `vect(1, ZeroTangent(), 3)` or `hcat([1,2], ZeroTangent())`, this materialises
+each zero `xdot` to be `zero(x)`.
+"""
+_make_real_zeros(xdots, xs) = map(_real_zero, xdots, xs)
+_real_zero(xdot, x) = xdot
+_real_zero(xdot::AbstractZero, x) = zero(x)
+
+# Fast paths. Should it also collapse all-Zero cases?
+_make_real_zeros(xdots::NTuple{<:Any, <:Number}, xs) = xdots
+_make_real_zeros(xdots::AbstractArray{<:Number}, xs) = xdots
+_make_real_zeros(xdots::AbstractArray{<:AbstractArray}, xs) = xdots
+
+frrule((_, xdd, _), ::typeof(_make_real_zeros), xdots, xs) = _make_real_zeros(xdots, xs), xdd  # not very sure!
+
+rrule(::typeof(_make_real_zeros), xdots, xs) = _make_real_zeros(xdots, xs), dydots -> (NoTangent(), dydots, NoTangent())
```

Comment:
Not so happy with these. The function is mathematically the identity, it just materialises some types, so perhaps having trivial rules is correct; the question is only whether they should also materialise (or destroy) some types. 

Their tests didn't work, https://github.com/JuliaDiff/ChainRules.jl/pull/565/files#diff-14d216cb53721d9360bbbcc399d93fff67f2594bd388162abbbe82c59ad1ba9fR44 , I think due to bugs elsewhere.

Could remove for now. The goal was eventually to write better 2nd-order tests in Diffractor.

Reply:
yeah might be safest to remove til we are happy with them.
(Can always type pirate them into Diffractor while debugging)

Reply:
Removed.